File: 2035-count-sub-islands/2035-count-sub-islands.cpp
#pragma GCC optimize("03", "unroll-loops");
static const int __ = [](){
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
    return 0;
}();

class Solution {
private:
    int m;
    int n;
    bool dfs(int r, int c, vector<vector<int>>& grid1, vector<vector<int>>& grid2) {
        if(r < 0 || r >= m || c < 0 || c >= n || grid2[r][c] == 0) return true;
        bool flag = grid1[r][c] == 1;
        grid2[r][c] = 0;
        flag &= dfs(r-1, c, grid1, grid2);
        flag &= dfs(r+1, c, grid1, grid2);
        flag &= dfs(r, c-1, grid1, grid2);
        flag &= dfs(r, c+1, grid1, grid2);


        return flag;
    }
public:
    int countSubIslands(vector<vector<int>>& grid1, vector<vector<int>>& grid2) {
        m = grid1.size();
        n = grid1[0].size();
        int ans = 0;
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(grid1[i][j] && grid2[i][j]) {
                    if(dfs(i, j, grid1, grid2)) ans++;
                }
            }
        }
        
        return ans;
    }
};

File: 2035-count-sub-islands/2035-count-sub-islands.go
var m, n int

func dfs(r, c int, grid1, grid2 [][]int) bool {
	if r < 0 || r >= m || c < 0 || c >= n || grid2[r][c] == 0 {
		return true
	}
	flag := grid1[r][c] == 1
	grid2[r][c] = 0
	flag = dfs(r-1, c, grid1, grid2) && flag
	flag = dfs(r+1, c, grid1, grid2) && flag
	flag = dfs(r, c-1, grid1, grid2) && flag
	flag = dfs(r, c+1, grid1, grid2) && flag

	return flag
}

func countSubIslands(grid1, grid2 [][]int) int {
	m = len(grid1)
	n = len(grid1[0])
	ans := 0
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid1[i][j] == 1 && grid2[i][j] == 1 {
				if dfs(i, j, grid1, grid2) {
					ans++
				}
			}
		}
	}
	return ans
}